Add --watch mode to `flutter test`

gamebox opened the pull request:
When flag is provided watch application directory for file changes and re-run tests on change, only incrementally compiling changed files. This provides a significant speedup on incremental edit/test cycle.

gamebox committed:
Merge pull request #1 from flutter/master

Catch up to flutter/flutter master

gamebox committed:
Merge pull request #2 from flutter/master

Merge from flutter/flutter

devoncarew committed:
rev the min dart sdk dep in the templates to 2.1.0 (#26235)

Commit:
Revert "Replace netls and netaddr with dev_finder" (#26249)

* Revert "Replace netls and netaddr with dev_finder (#26090)"

This reverts commit eee154affbce8a4d69ee4dbfed332edc7550d264.

Commit:
Force DatePicker value to stay within firstDate and lastDate upon year change (#26239)

* Force DatePicker value to stay within firstDate and lastDate upon year change

* updated per review feedback

goderbauer committed:
Remove long-deprecated TwoLevelList (#26238)

bkonyi committed:
Fixed typo (#26260)

bkonyi committed:
Manual engine roll with fixes: (#26252)

https://github.com/flutter/engine/compare/7112b72cc229...e5ec3cf3ea5c

git log
7112b72cc229e05d36716c3d7739885d3ffa72e6..e5ec3cf3ea5ce78c6c3116271601f60d30a799d4
--no-merges --oneline
e5ec3cf3e Dart SDK roll for 2019-01-08
08c95d27a Roll src/third_party/skia 55ff5d3ba881..1337f5b85978 (10
commits) (flutter/engine#7407)
e385f5cbb Roll src/third_party/skia 26d173fee72b..55ff5d3ba881 (12
commits) (flutter/engine#7406)
0f8273b0c Dart SDK roll for 2019-01-07
4036b260a Reset ParagraphBuilder after build() (flutter/engine#7401)
4820cbec4 Dart SDK roll for 2019-01-07
8eccb860d Add onStart hook to FlutterFragmentActivity
(flutter/engine#6719)
f2ea838b3 Roll src/third_party/skia b2fdcbf3033f..26d173fee72b (10
commits) (flutter/engine#7400)
5ca8aadaa Announce in/out of list (flutter/engine#6918)
4487d392d Replace Java code with equivalent, more concise code.
(flutter/engine#7398)
395b7852d Roll src/third_party/skia 46ee3f7a8ff5..b2fdcbf3033f (11
commits) (flutter/engine#7394)
5965f9084 Make `ParagraphConstraints` have const constructor
(flutter/engine#7346)
e02dd416a Roll src/third_party/skia a47eb455360f..46ee3f7a8ff5 (2
commits) (flutter/engine#7390)
f0038b3cc Roll src/third_party/skia 3ac3a4053f86..a47eb455360f (2
commits) (flutter/engine#7389)

kangwang1988 committed:
Improve the intergrity checking for "gradle wrapper". (#26069)

dnfield committed:
Fix red tree (Android module) (#26270)

dnfield committed:
Add compileOptions to android_host_app (#26271)

DanTup committed:
Add source line to snippet metadata file (#26274)

* Add source line to snippet metadata

* Parse sourceLine as an int

kangwang1988 committed:
Fix recursive link resulted `flutter doctor` stucking. See : #25885 and #25870 (#26017)

Commit:
Reland: Switch to dev_finder (#26250)

Commit:
Add polling module discovery for Fuchsia (#24994)

jxltom committed:
Do not ignore pubspec.lock in project templates (#24209)

* Remove *.lock from gitignore

* Remove pubspec.lock from gitignore

gamebox committed:
Pull test compiler out of FlutterPlatform

So that it can instead be used inside of `test/runner.dart`

gamebox committed:
Pull test bootstrap generation out of FlutterPlatform

So that it can be used in 'test/runner.dart'

jslavitz committed:
Teach drag start behaviors to DragGestureRecognizer (#26246)

* the onStart callback will report the location of the pointer where it wins the gesture arena by default instead of the pointer down location. Fixes all tests related to changing this default value.

seanfreiburg committed:
Fix flutter root error message string interpolation (#24944)

The variable isn't interpolating because it's using "${}" when ruby uses "#{}".

gamebox committed:
Move test compilation to test runner from flutter_platform

To allow for test watching and incremental compilation

kongshanshan33 committed:
Fix project directory has spaces lead to compile error when use 'flutter create -t module'command (#23817)

matthew-carroll committed:
Add consumedScrollExtent to SliverConstraints as reported by Viewport (#24551)

gamebox committed:
Add --watch flag to flutter test

Allows for incremental compile and running of test suites

gamebox commented:
@jonahwilliams Updated with a much more readable series of commits.

dnfield committed:
Check for duplicative Flutter.framework emeddings when building for Xcode 10 (#23188)

* Check for bad configuration in Xcode10

dnfield committed:
Update docs for editable_text (TextField) (#26269)

* update docs for editable_text (TextField)

* typo, clarify behavior when lines > 1

awdavies committed:
[flutter_driver] Move Fuchsia logging code. (#26266)

When setting the Fuchsia logging function, it should happen before any
initialization code, as init can still cause warning/error/info messages
to get printed to logs. Since the default stderr/stdout fd's aren't
correct, this can cause a program to crash for unclear reasons.

Commit:
Add link to Flutter brand guidelines (#26257)

marcoms committed:
Add new curve animations to class documentation (#26290)

jonahwilliams commented:
I haven't forgotten about this! Taking a longer look at it today.

gamebox committed:
Update runner for upstream changes to test_core API

Use new runTests method, and handle cleanup manually
when tests complete.

This is blocked by https://github.com/dart-lang/test/pull/973 being
merged and the flutter/flutter repo updating its
dependency on a version that includes that PR

gamebox commented:
The test failures are expected due to the facts laid out in the commit message of the commit just added

jonahwilliams commented:
Refactoring is fine but not necessary to this PR.

I pulled your gamebox/flutter and pointed test at gamebox/test with dependency overrides. `flutter test --watch` ran correctly the first time, but it did not detect any changes to the test directory or lib. I also could not exit the process without CTRL+C.

What is the intended workflow here? am I watching everything, a specific test file, or a library file?. I can see different cases where a watch command is useful depending on what I am working on:

- I am adjusting a library file and I want to rerun one test, multiple tests, multiple test files on change
- I am adjusting a test file and I want to rerun it on change.

Have you considered any other approaches? For example, I _could_ wire up either command myself with some shell commands, but it would be too slow. Thus it seems like caching the kernel files to disk might also speed up incremental runs. Of course, you would have to be more careful with invalidation!

apaatsio commented:
I'll just mention https://github.com/flutter/flutter/issues/4719 here since it's relevant.

gamebox commented:
Thanks @apaatsio , I'll check in with @Hixie about the operational semantics of this feature.

gamebox commented:
@jonahwilliams Just an update, my change dart-lang/test was approved.  Once it's merged, @grouma will cut a new release, and then I can update this branch to use it

gamebox committed:
Updates for new dart-lang/test release

Also, a little polish on the UX, following hot reload
semantics.

gamebox commented:
Alright @jonahwilliams , this includes the dart-lang/test updates, so this should reflect generally what I would like to release as a WIP feature in the tool.  The semantics have been changed from rerun-on-save, to the same as hot reload('r' to rerun, 'q' to quit).

jonahwilliams commented on packages/flutter_tools/pubspec.lock:
```diff
@@ -0,0 +1,502 @@
+# Generated by pub
```
Make sure to revert this file

gamebox committed:
Update test dependency

gamebox committed:
Delete pubspec.lock for flutter_tools

jonahwilliams commented:
Got everything working locally, this is pretty neato!

jonahwilliams commented on packages/flutter_tools/lib/src/test/runner.dart:
```diff
@@ -68,19 +72,51 @@ Future<int> runTests(
-    updateGoldens: updateGoldens,
-    projectRootDirectory: fs.currentDirectory.uri,
-  );
+  final Uri projectRootDirectory = fs.currentDirectory.uri;
```
Consider removing some of the empty lines here

gamebox replied:
Done

jonahwilliams commented on packages/flutter_tools/lib/src/test/runner.dart:
```diff
@@ -89,14 +125,65 @@ Future<int> runTests(
+  
+  final Function printReloadMessage = () {
+    printStatus(
+      "Press 'r' to hot reload your tests, 'q' to quit",
```
Instead of hot reload call this rerunning, it is more descriptive.

gamebox replied:
Done

jonahwilliams commented on packages/flutter_tools/lib/src/test/runner.dart:
```diff
@@ -89,14 +125,65 @@ Future<int> runTests(
+      terminal.onCharInput.asBroadcastStream().listen((String char) async {
+        switch(char) {
+          case 'r':
+            stdout.writeln(terminal.clearScreen());
```
I really don't like clearing the terminal here, I lose the context of my previous test run/failure.

gamebox replied:
Removed, I thought it was cleaner to see what the current test state is while running.

jonahwilliams commented on packages/flutter_tools/lib/src/test/runner.dart:
```diff
@@ -89,14 +125,65 @@ Future<int> runTests(
+        switch(char) {
+          case 'r':
+            stdout.writeln(terminal.clearScreen());
+            printStatus('Recompiling test files...\n', emphasis: true, color: TerminalColor.blue);
```
I would use a Status instead:
```
Status status = logger.startProgress('Recompiling test files...', timeout: null);
...
status.stop();

```

gamebox replied:
Didn't see this.  This is awesome.  Done

jonahwilliams commented on packages/flutter_tools/lib/src/test/runner.dart:
```diff
@@ -89,14 +125,65 @@ Future<int> runTests(
+            completer.complete();
+        }
+      }, onDone: completer.complete);
+
```
Consider removing some of these empty lines

jonahwilliams commented on packages/flutter_tools/lib/src/test/runner.dart:
```diff
@@ -89,14 +125,65 @@ Future<int> runTests(
+    try {
+      test.completeShutdown();
+    } on StateError catch (e) {
+      printTrace(e.toString());
```
printError ?

gamebox replied:
Of course, Done.

jonahwilliams commented on packages/flutter_tools/lib/src/commands/test.dart:
```diff
@@ -124,6 +132,10 @@ class TestCommand extends FlutterCommand {
     if (files.isEmpty) {
+      if (argResults['watch']) {
+        throwToolExit(
+          'When using --watch, you must specify one or more test file(s) to run.');
```
Presumably the user could manually specify every single file right? Might as well make that supported fully

gamebox replied:
The philosophy here is to make the main use case easy, and what we just allow to be done, but with some difficulty.  A sensible alternative would be to restrict to a single test file.  I don't have any _strong_ opinions either way, but we are solving for taking a single - or a small number of - test(s) from red to green.

jonahwilliams replied:
I agree in principle, but there is an edge case where the behavior is painful enough that we may as well allow it: the project as one or few tests files

gamebox replied:
Agreed offline to remove this stricture

jonahwilliams commented on packages/flutter_tools/lib/src/commands/test.dart:
```diff
@@ -167,14 +182,18 @@ class TestCommand extends FlutterCommand {
     if (collector != null) {
-      if (!await collector.collectCoverageData(
-          argResults['coverage-path'], mergeCoverageData: argResults['merge-coverage']))
+      final bool coverageDataCollected = await collector.collectCoverageData(
```
Revert this change

gamebox replied:
Done

gamebox replied:
Done

jonahwilliams commented on packages/flutter_tools/lib/src/commands/test.dart:
```diff
@@ -139,7 +151,8 @@ class TestCommand extends FlutterCommand {
 
     CoverageCollector collector;
-    if (argResults['coverage'] || argResults['merge-coverage']) {
+    if (argResults['coverage'] ||
```
Revert this change

gamebox replied:
Done

jonahwilliams commented on packages/flutter_tools/lib/src/commands/test.dart:
```diff
@@ -75,6 +75,14 @@ class TestCommand extends FlutterCommand {
+        defaultsTo: false,
+        negatable: false,
+        hide: true, // TODO(gamebox): Remove when UX is finalized
+        help: 'After initial run, continue watching application directory for '
```
"similar semantics to hot reload" - is that actually true? do constants not change between test runs now?

gamebox replied:
Sorry, I'm not sure what you mean by "do constants not change between test runs now?"  We can discuss offline.  By "similar semantics", I meant we only recompile the changed files and then rerun the test code - not to mention that the rerunning happens by pressing 'r' in the terminal/stdin.  Which is why I did not say "the exact same semantics".

gamebox replied:
Changed the language to reduce confusion.

jonahwilliams commented on packages/flutter_tools/lib/src/commands/test.dart:
```diff
@@ -75,6 +75,14 @@ class TestCommand extends FlutterCommand {
         help: 'Whether matchesGoldenFile() calls within your test methods should '
               'update the golden files rather than test for an existing match.',
       )
+      ..addFlag('watch',
```
Even though there is some file watching going on, I don't think that is the defining feature of this. Maybe "interactive?" I'm not quite sure.

gamebox replied:
I'm open to either, I have it hidden so that we can feel free to change this in the near future.

gamebox replied:
Leaving as is until someone has a better name for the flag.  Since the files are in fact being watched, I think the current name makes sense.

gamebox commented:
Will address the rest of the comments in the morning.  Thanks for the thorough review.  Most of my issues remaining have to do with updating the dependencies.  Will discuss offline.

gamebox commented on packages/flutter_tools/lib/src/commands/test.dart:
```diff
@@ -124,6 +132,10 @@ class TestCommand extends FlutterCommand {
 
     Directory workDir;
     if (files.isEmpty) {
+      if (argResults['watch']) {
```
Done

gamebox commented on packages/flutter_tools/lib/src/test/bootstrap.dart:
```diff
@@ -0,0 +1,181 @@
+      host: host,
+      updateGoldens: updateGoldens,
+    );
+  }
```
Done

gamebox commented on packages/flutter_tools/lib/src/test/bootstrap.dart:
```diff
@@ -0,0 +1,181 @@
+import 'package:meta/meta.dart';
```
Done

gamebox committed:
Address review feedback

Removed references to 'hot reload', and cleaned style nits.
Also, updated packages to pass analyze.

gamebox committed:
Import from path, not flutter_tools package

gamebox committed:
Remove trailing whitespace

gamebox committed:
Update test template

gamebox committed:
FREEZE.unindexed

gamebox commented:
That last "queued" test has actually completed and passed.  Don't know if that's something wrong on Cirrus or GH's side.

gamebox commented:
Closing in favor of (#27192)
